types/wlr_surface.c

```diff
@@ -185,26 +188,23 @@ static void surface_update_damage(pixman_region32_t *buffer_damage,
 			prev_y = tmp;
 		}
 
-		pixman_region32_union_rect(buffer_damage, buffer_damage,
-			prev_x * previous->scale, prev_y * previous->scale,
-			previous->buffer_width, previous->buffer_height);
-		pixman_region32_union_rect(buffer_damage, buffer_damage, 0, 0,
-			current->buffer_width, current->buffer_height);
+		pixman_region32_union_rect(&surface_damage, &surface_damage, prev_x,
```

Comment:
I'd prefer using buffer coordinates here.

Reply:
That's the source of the bug (in `weston-fullscreen` resizing when transformed). If transformed, the buffer's origin and the surface's origin don't match, so the "overflowing" damage here doesn't correspond to where the surface was previously.
TBH, I don't agree with setting the damage beyond the bounds of the surface here at all: it breaks the implicit assumption of the damage being limited by the surface. IMO, overflowing resizing damage should be handled by the compostor/rootston.

Reply:
We expose both the original damage (via `surface->current`) and the easy-to-use damage. It's up to the compositor to manage it itself.

There's no assumption at all about the damage being inside of the buffer bounds. Maybe we should make this more explicit in the header docs.

Reply:
Fair enough. In any case, buffer coordinates really aren't feasible here, since they'd have to be transformed in exactly the same way the surface coordinates are, and that's redundant effort.

Reply:
Hmm, what do you mean? We can just use `current->buffer_width` and `current->buffer_height` as it was done previously. No need to use `current->width` and `current->height`.

Reply:
Let me take the easiest case: flipped. Suppose originally the surface and buffer was 20x10, with the upper left corner at (10,0). Say someone resized the surface/buffer to 10x10.

If we use `current->buffer_*` here, the buffer damage is ((0,0),(20,10)), which on flipping with buffer size 10x10 (when getting back the surface damage during output damage calculation) gives the damage on the output as (10,0) + ((-10,0),(10,10)) == ((0,0),(20,10)) (translation), which is wrong: it should be ((10,0),(30,10)), the previous surface bounds.